op-chain-ops/ether/migrate.go

```diff
@@ -27,73 +45,348 @@ var (
+// It gets passed via channel to the collector process.
+type accountData struct {
+	balance    *big.Int
+	legacySlot common.Hash
+	address    common.Address
+}
 
-func MigrateLegacyETH(db *state.StateDB, addresses FilteredOVMETHAddresses, chainID int, noCheck bool) error {
+type DBFactory func() (*state.StateDB, error)
+
+// MigrateBalances migrates all balances in the LegacyERC20ETH contract into state. It performs checks
+// in parallel with mutations in order to reduce overall migration time.
+func MigrateBalances(mutableDB *state.StateDB, dbFactory DBFactory, addresses []common.Address, allowances []*crossdomain.Allowance, chainID int, noCheck bool) error {
 	// Chain params to use for integrity checking.
 	params := crossdomain.ParamsByChainID[chainID]
 	if params == nil {
 		return fmt.Errorf("no chain params for %d", chainID)
 	}
 
-	// Log the chain params for debugging purposes.
-	log.Info("Chain params", "chain-id", chainID, "supply-delta", params.ExpectedSupplyDelta)
+	return doMigration(mutableDB, dbFactory, addresses, allowances, params.ExpectedSupplyDelta, noCheck)
+}
+
+func doMigration(mutableDB *state.StateDB, dbFactory DBFactory, addresses []common.Address, allowances []*crossdomain.Allowance, expDiff *big.Int, noCheck bool) error {
+	// We'll need to maintain a list of all addresses that we've seen along with all of the storage
+	// slots based on the witness data.
+	slotsAddrs := make(map[common.Hash]common.Address)
+	slotsInp := make(map[common.Hash]int)
 
-	// Migrate the legacy ETH to ETH.
-	log.Info("Migrating legacy ETH to ETH", "num-accounts", len(addresses))
-	totalMigrated := new(big.Int)
-	logAccountProgress := util.ProgressLogger(1000, "imported accounts")
+	// For each known address, compute its balance key and add it to the list of addresses.
+	// Mint events are instrumented as regular ETH events in the witness data, so we no longer
+	// need to iterate over mint events during the migration.
 	for _, addr := range addresses {
-		// Balances are pre-checked not have any balances in state.
+		sk := CalcOVMETHStorageKey(addr)
+		slotsAddrs[sk] = addr
+		slotsInp[sk] = BalanceSlot
+	}
+
+	// For each known allowance, compute its storage key and add it to the list of addresses.
+	for _, allowance := range allowances {
+		sk := CalcAllowanceStorageKey(allowance.From, allowance.To)
+		slotsAddrs[sk] = allowance.From
+		slotsInp[sk] = AllowanceSlot
+	}
+
+	// Add the old SequencerEntrypoint because someone sent it ETH a long time ago and it has a
+	// balance but none of our instrumentation could easily find it. Special case.
+	entrySK := CalcOVMETHStorageKey(sequencerEntrypointAddr)
+	slotsAddrs[entrySK] = sequencerEntrypointAddr
+	slotsInp[entrySK] = BalanceSlot
+
+	// WaitGroup to wait on each iteration job to finish.
+	var wg sync.WaitGroup
+	// Channel to receive storage slot keys and values from each iteration job.
+	outCh := make(chan accountData)
```

Comment:
An unbuffered channel is fine here b/c we aren't able to deadlock in this code, but I'd lean to a buffered channel here b/c we don't need the synchronization. IDK if it will have any performance impact though.

Reply:
We do need the synchronization. There's nothing explicitly draining `outCh`, so if we make this channel buffered we'll have to drain it after waiting for the all the workers to exit. Keeping this channel unbuffered guarantees that the account data channel is empty before the `wg.Wait()` unblocks.

Reply:
Oh I see, the collector is it's own go-routine. It doesn't seem like it need to be, but making it buffered does require a fair bit more changes.

Reply:
Yeah. LMK if you feel strongly about any modifications here, otherwise I'll leave as-is.